pandas/core/indexes/period.py

```diff
@@ -148,144 +177,176 @@ class PeriodIndex(PeriodArrayMixin, DatelikeOps, DatetimeIndexOpsMixin,
+        msg = ('Setting {cls}.freq has been deprecated and will be '
+               'removed in a future version; use {cls}.asfreq instead. '
+               'The {cls}.freq setter is not guaranteed to work.')
+        warnings.warn(msg.format(cls=type(self).__name__),
+                      FutureWarning, stacklevel=2)
+        # PeriodArray._freq isn't actually mutable. We set the private _freq
+        # here, but people shouldn't be doing this anyway.
+        self._data._freq = value
+
+    # ------------------------------------------------------------------------
+    # Index Constructors
 
     @classmethod
     def _simple_new(cls, values, name=None, freq=None, **kwargs):
+        # type: (PeriodArray, Any, Any) -> PeriodIndex
         """
         Values can be any type that can be coerced to Periods.
         Ordinals in an ndarray are fastpath-ed to `_from_ordinals`
         """
-        if not is_integer_dtype(values):
-            values = np.array(values, copy=False)
-            if len(values) > 0 and is_float_dtype(values):
-                raise TypeError("PeriodIndex can't take floats")
-            return cls(values, name=name, freq=freq, **kwargs)
-
-        return cls._from_ordinals(values, name, freq, **kwargs)
+        assert isinstance(values, PeriodArray)
+        result = object.__new__(cls)
+        result._data = values
+        result.name = name
+        result._reset_identity()
+        return result
 
     @classmethod
     def _from_ordinals(cls, values, name=None, freq=None, **kwargs):
         """
         Values should be int ordinals
         `__new__` & `_simple_new` cooerce to ordinals and call this method
         """
-        result = super(PeriodIndex, cls)._from_ordinals(values, freq)
-
-        result.name = name
-        result._reset_identity()
+        data = PeriodArray._from_ordinals(values, freq=freq)
+        result = cls._simple_new(data, name=name)
         return result
 
-    def _shallow_copy_with_infer(self, values, **kwargs):
+    def _shallow_copy(self, values=None, **kwargs):
+        # TODO: simplify, figure out type of values
+        if values is None:
+            values = self._values
+        if not isinstance(values, PeriodArray):
+            if (isinstance(values, np.ndarray) and
+                    is_integer_dtype(values.dtype)):
+                values = PeriodArray._from_ordinals(values, freq=self.freq)
+            else:
+                # in particular, I would like to avoid complex_new here.
+                # Some people seem to be calling use with unexpected types
+                # Index.difference -> ndarray[Period]
+                # DatetimelikeIndexOpsMixin.repeat -> ndarray[ordinal]
+                # I think that once all of Datetime* are EAs, we can simplify
+                # this quite a bit.
+                values = PeriodArray._complex_new(values, freq=self.freq)
+
+        # I don't like overloading shallow_copy with freq changes.
```

Comment:
Yah, I'm not wild about this.  I still think the best option is to nail down the constructors before doing the whole PeriodArray changeover.